### ISSUE TITLE
combine --reload and --prefetch into --load

### ISSUE THREAD

**Issue by ry** [denoland/deno#1582]
or rename --prefetch to --preload ? (to better correspond to --reload)

**Comment by hayd** [comment]
I think there is a valid usecase for these being distinct. In a Dockerfile you may want to do:

```dockerfile
ADD deps.ts                  # layer only re-run if deps.ts changes 
RUN deno --prefetch deps.ts  # layer only re-run if deps.ts changes 
ADD .
RUN deno --prefetch main.ts  # doesn't fetch deps.ts since they're already cached 
```
Specifically you don't want either of these runs to execute, and in the latter you don't want it to clear the deps.ts cache.

**Comment by ry** [comment]
I think you have a typo ?

**Comment by hayd** [comment]
Neither of these are --reload. Is that what you mean?

**Comment by ry** [comment]
Yeah. I don't understand.

The idea is that --load would be exactly as prefetch is now. In particular it doesn't run evaluate the program. So people who want to reload will have to do two commands: `deno $module --load && deno $module`
I think this decently ergonomic. I'm striving to keep the CLI very simple - the unix philosophy is that each command should just do one thing - I think combining these flags goes in that direction.

**Comment by hayd** [comment]
Example above it's desired to load but not reload (not clear the cache).

> the unix philosophy is that each command should just do one thing 

Doesn't that suggest one for emptying cache and one for loading?

Perhaps these should be subcommands e.g. `deno load mod.ts` and `deno clear`?

**Comment by ry** [comment]
i mean clearing the cache is just rm -rf ~/.deno/deps - i dont think we need a special command for that.

**Comment by ry** [comment]
regarding subcommands - let's stick with flags for now. flags are more ergonomic imo

**Comment by hayd** [comment]
RE subcommands, I think a good rule of thumb would be "if it runs the script but changes how it behaves it's a flag, if it does something else it ought to be a subcommand".

e.g. 
- --refresh, clears cache and runs script is a flag.
- --allow-write, allows write when it runs script is a flag.
- load, populates cache but doesn't run script is subcommand.
- format, runs prettier on file is subcommand.

etc. In other issues you've suggested subcommands would exist so I don't understand the positioning the cursor concern.

**Comment by ry** [comment]
I'm not against subcommand necessarily but I'd rather push one means of specifying options until it becomes unreasonable. We already have the flag parsing infrastructure in place - so it's easy to plod along in that direction a bit longer.

> "if it runs the script but changes how it behaves it's a flag, if it does something else it ought to be a subcommand".

I think that's a good and simple heuristic - but it's more complicated for users than "everything is a flag"

>  In other issues you've suggested subcommands would exist so I don't understand the positioning the cursor concern.

I'm open to subcommands - just skeptical and not eager to complicate the CLI.

**Comment by jedahan** [comment]
Another useful hueristic is "could someone combine valid flags in an invalid way" i.e. would scoping flags to a subcommand prevent conflicts of intent. 

tldr is [developing a client interface specification](https://github.com/tldr-pages/tldr/pull/2706#issuecomment-456994303) and one complication is the overloading of -- for both actions and options, and this discussion seemed similar.

**Comment by sholladay** [comment]
Just my two cents:
 - I don't like `deno load --reload`, it looks very strange to me when I read it and is unclear what exactly `load` is doing that `--reload` isn't doing.
 - I like the existing `--prefetch` and `--reload` option names, and I like `--preload` too, but I think `--preload` and `--reload` are _too_ similar. They are only one letter apart, very prone to typos and brain hiccups.
 - If succinctness is one of the goals here, `--reload` could be renamed to `--fresh`. It's shorter and also isn't prone to the "too similar to reload" problem.
 - Maybe consider having `--prefetch` imply `--reload` automatically. After all, the idea is to go get dependencies from the network. I'm skeptical that there are many situations where you want to prefetch dependencies with some of them being potentially stale. Seems simpler to me to have an all or nothing approach where you choose between "bring your own dependencies" or prefetch them fresh.

**Comment by hayd** [comment]
> Maybe consider having --prefetch imply --reload 

This is precisely what I was arguing against, with the example above.

**Comment by ry** [comment]
> This is precisely what I was arguing against, with the example above.

I still don't really understand your use-case. Let me see if I understand this correctly: you have a program with some source files that have been prefetched. And in one of them you've added a new import. You want the ability to prefetch that new import without reloading all of the existing cached files. Does this describe what you're looking for?

**Comment by sh7dm** [comment]
I'd like to work on it

**Comment by hayd** [comment]
Not quite, the dependencies don't change frequently but the app code does, this is specifically useful in leveraging the docker layering in a development cycle.

**Comment by ry** [comment]
@hayd Ok got it... Not sure I agree this necessitates a separate flag - if you're building a docker container seems reasonable you can just re-download the deps... Browsers don't provide granulated hard reloads - you have to hard reload everything. But let me think about it.

**Comment by hayd** [comment]
perhaps --compile is a better phrasing that --prefetch ?

**Comment by bartlomieju** [comment]
CC @ry, no longer valid with combined `--reload` and `--prefetch` flags?
